#517157 [C#テキスト]4章演習問題

sugisaka-being opened the pull request:
4章演習問題の解答を作成しました。

4章の概要[基本イディオム]
・初期化に関するイディオム
・判定と分岐に関するイディオム
・繰り返しのイディオム
・条件演算子、null合体演算子によるイディオム
・プロパティに関するイディオム
・メソッドに関するイディオム
・その他のイディオム

sugisaka-being committed:
#517157 [C#テキスト]4章演習問題

演習問題の解答を作成しました。

teshima-being commented on Chapter4/Chapter4-1-1/YearMonth.cs:
```diff
@@ -0,0 +1,52 @@
+        public int Month { get; }
+
+        // 2.
+        /// <summary>
+        /// 21世紀
+        /// </summary>
+        public bool Is21Century => 2001 <= Year && Year <= 2100;
```
ラムダ式を活用できていて良いですね！
あとはクラス内のプロパティを利用しているので`Year`に`this`を付けておきましょう。
同じような修正が必要な個所がほかにもないか確認してみてください。

sugisaka-being replied:
コメントありがとうございます。
YearMonthクラスにおいて、クラス内のプロパティを利用している部分全てに`this`が付くよう、修正いたしました。

teshima-being commented on Chapter4/Chapter4-1-1/YearMonth.cs:
```diff
@@ -0,0 +1,52 @@
+        /// 1カ月後を求めるメソッド
+        /// </summary>
+        /// <returns>1カ月後の年月</returns>
+        public YearMonth AddOneMonth() {
+            if (Month == 12)
+                return new YearMonth(Year + 1 , 1);
+            return new YearMonth(Year, Month + 1);
+        }
```
if文の`{}`を省略するときは前後の処理も含めて注意してください。
今回の場合if文が`true`の場合に次の行ですぐに`return`しているので期待した動作になってはいますが、`{}`を省略していることにより、のちのちコードに修正を加えた際にバグが発生しやすい状態になっています。
```
        public YearMonth AddOneMonth() {
            if (Month == 12) 
                Console.WriteLine("年が変わりました。");　← 1行追加
                return new YearMonth(Year + 1 , 1);　← if文の条件判定に関係なく常に実行される。
            return new YearMonth(Year, Month + 1);　← 上のreturnが常に実行されるため到達できない。
        }
```

こういったことを防ぐために、{}を省略する場合は下記のように条件がtrueの場合の処理も含めて1行で記述する（①）か、そもそも{}を省略せずに記載する（②）ほうが良いです。
可読性が大きく下がらないのであれば三項演算子を活用する手もあります。

①
```
        public YearMonth AddOneMonth() {
            if (Month == 12) return new YearMonth(Year + 1, 1);
            return new YearMonth(Year, Month + 1);
        }
```
②
```
        public YearMonth AddOneMonth() {
            if (Month == 12) {
                return new YearMonth(Year + 1 , 1);
            }
            return new YearMonth(Year, Month + 1);
        }
```

sugisaka-being replied:
コメントありがとうございます。
省略できる部分は全て省略した方が良いというわけではなく、後にコードを修正する、追加する可能性があることを考えて、バグが起きにくいように記述していくことが大切なのですね。非常に勉強になります。
今回ご指摘いただいた部分は、{}を省略せずに記載する（②）のコードへと修正を行いました。

teshima-being commented on Chapter4/Chapter4-1-1/YearMonth.cs:
```diff
@@ -0,0 +1,52 @@
+        /// ToStringをオーバーライドして年月を表示するメソッド
+        /// </summary>
+        /// <returns>〇年〇月</returns>
+        public override string ToString() {
+            return $"{Year}年{Month}月";
+        }
```
メソッドについてもラムダ式を活用することで、`{}` や` return` を省略して簡潔に記述することが出来ます。
処理が1行で済むようであれば積極的に利用してみましょう。
例）`public override string ToString() => $"{this.Year}年{this.Month}月";`

thisも忘れないようにしましょう。

sugisaka-being replied:
コメントありがとうございます。
メソッドにおいてもラムダ式が活用できること、初めて知りました。とても勉強になります。

` public override string ToString() => $"{this.Year}年{this.Month}月";`へとコードを修正いたしました。

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,68 @@
+            };
+
+            // 2.
+            foreach (var wYearMonth in wYearMonthCollection) {
+                Console.WriteLine($"{wYearMonth.Year}年{wYearMonth.Month}月");
```
「〇年〇月」と表示する方法として間違っていませんが、今回は問題にてYearMonthクラスにToStringメソッドをオーバーライドしたものが実装されていると思います。
せっかくなのでそれを利用するようにしてみましょう。

sugisaka-being replied:
コメントありがとうございます。
以下のようにコードを修正いたしました。
```
foreach (var wYearMonth in wYearMonthCollection) {
    Console.WriteLine(wYearMonth.AddOneMonth());
}
```

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,68 @@
+        /// </summary>
+        /// <param name="vYearMonths">年月</param>
+        /// <returns>配列の最初の21世紀（無い場合はnull）</returns>
+        static YearMonth FindFirst21Century(YearMonth[] vYearMonths) {
+            foreach (var wYearMonth in vYearMonths) {
+                if (wYearMonth.Is21Century)
+                    return wYearMonth;
+            }
+            return null;
+        }
```
動きとしては問題ないですが 2点だけ。

(1)
メソッドの引数の型を`YearMonth[]`に指定していることで、再利用がしにくくなっています。
例えば` List<YearMonth>`などに対してはこのメソッドが使えません。
（問題1で定義している`YearMonth`の配列を`List<YearMonth>`に変えてみると分かると思います。）

メソッド内の処理に配列でないとできないものがある場合などは仕方ないですが、そうではないなら出来るだけ広い型に適用できるようにしておくと汎用性が高まります。
今回は`foreach`をしているだけなので、引数の型を`IEnumerable<YearMonth>`にしておきましょう。
また、この機にIEnumerable型についても調べてみてください。

(2)
※追加課題
問題文に`foreach`を使うように指定がありますが、練習として`foreach`を使わず実装してみてください。
「条件を満たすコレクションの最初の要素を返し、見つからなかった場合はデフォルト値（参照型の場合は`null`）を返す」便利なものがLINQにあったと思います。

sugisaka-being replied:
コメントありがとうございます。
(1)引数の型を`IEnumerable<YearMonth>`へと変更いたしました。
(2)追加課題の解答を作成しました。
```
static YearMonth FindFirst21CenturyAdd(IEnumerable<YearMonth> vYearMonths) {
    return vYearMonths.FirstOrDefault(x => x.Is21Century);
}
```

teshima-being replied:
FirstOrDefaultで書き換えられましたね！
今回は問題文で指定があったので、FindFirst21Centuryメソッドを定義してくれていましたが、実際にはFirstOrDefaultを利用すればメソッド自体不要になります。

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,68 @@
+            Console.WriteLine(FindFirst21Century(wYearMonthCollection)?.ToString() ?? "21世紀のデータはありません");
+
+            // 5.
+            var wOneMonthLaters = wYearMonthCollection.Select(x => x.AddOneMonth()).ToArray();
```
右辺から変数の型が明らかな場合に積極的に`var`を利用できていて良いですね！
（今回は.ToArray()しているため配列であることが自明）

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,68 @@
+        */
+        static void Main(string[] args) {
+            // 1.
+            var wYearMonthCollection = new YearMonth[] {
```
細かいですが、`wYearMonthCollection`は`wYearMonths`の方が簡潔で分かりやすいですね。
変数名は短ければ良いわけではありませんが、長くなっても情報量がほぼ変わらないのであれば、短く簡潔になっている方が読みやすいです。

sugisaka-being replied:
コメントありがとうございます。
`wYearMonthCollection`を`wYearMonths`へと修正いたしました。

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,68 @@
+                Console.WriteLine($"{wYearMonth.Year}年{wYearMonth.Month}月");
+            }
+
+            // 4.
+            Console.WriteLine(FindFirst21Century(wYearMonthCollection)?.ToString() ?? "21世紀のデータはありません");
```
三項演算子を活用できていますね！

問題文では「年を表示してください」とあるので、月は表示しないようにしてみましょう。
例）2050年

sugisaka-being replied:
コメントありがとうございます。
月が出力されないようにプログラムを修正いたしました。
`Console.WriteLine(FindFirst21Century(wYearMonths)?.Year + "年" ?? "21世紀のデータはありません");`

teshima-being replied:
`Console.WriteLine(FindFirst21Century(wYearMonths)?.Year + "年" ?? "21世紀のデータはありません");`
を実行したところ、21世紀のデータが存在しない場合に「年」と表示されました。
21世紀のデータが存在しない場合に「21世紀のデータはありません」と表示されるようにしてください。

teshima-being reviewed (commented):
テキストで学習したことを積極的に活用できていますね！
いくつか修正が必要な個所と、追加で考えて欲しいことについてコメントしましたので、確認お願いします。

sugisaka-being committed:
#517157 [C#テキスト]4章演習問題

クラス内のプロパティを利用しているものにthisを付けました。
if文において、省略していた{}を記載するようにしました。
メソッドにおいて、ラムダ式を活用しました。
ToStringメソッドを利用するコードを記述しました。
問題4-1-2の3.において、引数の型を変更しました。
追加課題の解答を作成しました。
変数名を簡潔にしました。
問題4-1-2の4.において、年のみを出力するようにしました。

teshima-being commented on Chapter4/Chapter4-1-2/Program4-1-2.cs:
```diff
@@ -0,0 +1,82 @@
+            };
+
+            // 2.
+            foreach (var wYearMonth in wYearMonths) {
+                Console.WriteLine(wYearMonth.AddOneMonth());
```
修正前と動作が変わってしまっているため、修正をお願いします。
問題2では配列YearMonthsの各要素をコンソールにそのまま表示するだけで、1か月後に変換する必要はないかと思います。
コードを修正したら、修正前後で意図しない挙動の変更が起きていないか確認することも重要ですね。

sugisaka-being replied:
コメントありがとうございます。
ToStringメソッドを利用するつもりが誤ってAddOneMonthメソッドを利用してしまっていました。
今後はエラーが出ないかだけではなく、動作内容までしっかりと確認を行っていきたいです。

teshima-being reviewed (commented):
1点だけ修正前と挙動が変わっている箇所があったため修正をお願いします。
練習のため意図的にそうしているのであれば、その旨をGitHubのコメントで書いていただければそれでもOKです。

sugisaka-being committed:
#517157 [C#テキスト]4章演習問題

利用するメソッドを変更いたしました。

m-hotta-being commented on Chapter4/Chapter4-1-1/YearMonth.cs:
```diff
@@ -0,0 +1,51 @@
+        /// 1カ月後を求めるメソッド
+        /// </summary>
+        /// <returns>1カ月後の年月</returns>
+        public YearMonth AddOneMonth() {
+            if (this.Month == 12) {
+                return new YearMonth(this.Year + 1, 1);
+            } 
+            return new YearMonth(this.Year, this.Month + 1);
+        }
```
【FYI: 情報共有。アクション不要です。】
よいわるいは別にして参考まで。
あと、ここに出てくる12月とか1月ってマジックナンバーですよね。
適切な名前を付けられるといいですね。
```suggestion
        public YearMonth AddOneMonth() =>
            this.Month == 最後の月 ? new YearMonth(this.Year + 1, 最初の月) : new YearMonth(this.Year, this.Month + 1);
```

sugisaka-being replied:
コメントありがとうございます。勉強になります。
ちょうど昨日、マジックナンバーを排除する方が良いケースを例にマジックナンバーについての知識を得ました。
今後は適切な名前を付けてコードを記述していきたいです。

m-hotta-being replied:
昔読んだ本に「1, 0, -1 以外はすべてマジックナンバーだ！」みたいな話がありました。
1, 0, -1 でもマジックナンバーのときはありますし、別に数字に限定した話でもないんですが、個人的には結構考え方の指針になってます。

sugisaka-being committed:
#517157 [C#テキスト]4章演習問題

「年」の表示する際、ToString()を利用する方法に修正をいたしました。

sugisaka-being committed:
#517157 [C#テキスト]4章演習問題

ToStringメソッドを利用するようにコードを修正しました。

teshima-being commented:
レビューOKです。4章お疲れ様でした。
